### PR TITLE
std: separate TLS key creation from TLS access

### PR DESCRIPTION
**Pull request by joboet**
Currently, `std` performs an atomic load to get the OS key on every access to `StaticKey` even when the key is already known. This PR thus replaces `StaticKey` with the platform-specific `get` and `set` function and a new `LazyKey` type that acts as a `LazyLock<Key>`, allowing the reuse of the retreived key for multiple accesses.

Related to #110897.

### CODE REVIEW

File: library/std/src/sys/thread_local/os.rs [hunk 1]
```diff
@@ -66,19 +66,19 @@ impl<T: 'static> Key<T> {
     /// The resulting pointer may not be used after reentrant inialialization
     /// or thread destruction has occurred.
     pub fn get(&'static self, i: Option<&mut Option<T>>, f: impl FnOnce() -> T) -> *const T {
-        // SAFETY: (FIXME: get should actually be safe)
-        let ptr = unsafe { self.os.get() as *mut Value<T> };
+        let key = self.key.force();
+        let ptr = unsafe { get(key) as *mut Value<T> };
```
Comment (jhpratt) [inline library/std/src/sys/thread_local/os.rs:70]: This & other instances in this file need safety comments. They didn't really have them previously, but we should keep the FIXME around if nothing else.

Comment (joboet) [reply library/std/src/sys/thread_local/os.rs:70]: I was the one to add the FIXME, but it's no longer right: the free `get` function relies on only receiving correct keys whereas the `get` method didn't even take an explicit key.

Sure, I'll add safety comments!

Comment (jhpratt) [reply library/std/src/sys/thread_local/os.rs:70]: Ah, okay! Even stating that is reasonable.

**Comment by joboet** [comment]
Thank you!
@bors r=@jhpratt

**Pull request merged**